src/target/source/codegen_params.cc

```diff
@@ -0,0 +1,214 @@
+  }
+
+  int elements_per_row = 16;
+  while (elements_per_row > 1 &&
+         (elements_per_row * one_element_size_bytes) > (kMaxLineLength - indent_chars)) {
+    elements_per_row /= 2;
+  }
+
+  std::string indent_str(indent_chars, ' ');
+  os << indent_str;
+
+  auto shape = arr.Shape();
+  int num_elements = 1;
+  for (auto shape_elem : shape) {
+    num_elements *= shape_elem;
+  }
+
+  std::unique_ptr<DLManagedTensor, DLManagedTensorDeleter> tensor(arr.ToDLPack());
+  auto old_fmtflags = os.flags();
+  os.setf(std::ios::internal | std::ios::hex,
+          std::ios::adjustfield | std::ios::basefield | std::ios::showbase);
+  os.fill('0');
+  switch (arr_type.code()) {
+    case runtime::DataType::kInt:
+      CHECK(arr_type.bits() == 8 || arr_type.bits() == 16 || arr_type.bits() == 32 ||
+            arr_type.bits() == 64)
+          << "CodegenParams: only support generating 8-, 16-, 32-, or 64-bit integer params; saw "
+          << arr_type.bits() << "-bit array";
+      if (arr_type.bits() == 8) {
+        PrintArray<int8_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str, os);
+      } else if (arr_type.bits() == 16) {
+        PrintArray<int16_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str, os);
+      } else if (arr_type.bits() == 32) {
+        PrintArray<int32_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str, os);
+      } else if (arr_type.bits() == 64) {
+        PrintArray<int64_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str, os);
+      } else {
+        CHECK(false) << "should not get here";
+      }
+      break;
+
+    case runtime::DataType::TypeCode::kUInt:
+      CHECK(arr_type.bits() == 8 || arr_type.bits() == 16 || arr_type.bits() == 32 ||
+            arr_type.bits() == 64)
+          << "CodegenParams: only support generating 8-, 16-, 32-, or 64-bit integer params; saw "
+          << arr_type.bits() << "-bit array";
+
+      if (arr_type.bits() == 8) {
+        PrintArray<uint8_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str, os);
+      } else if (arr_type.bits() == 16) {
+        PrintArray<uint16_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str,
+                             os);
+      } else if (arr_type.bits() == 32) {
+        PrintArray<uint32_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str,
+                             os);
+      } else if (arr_type.bits() == 64) {
+        PrintArray<uint64_t>(tensor->dl_tensor.data, num_elements, elements_per_row, indent_str,
+                             os);
+      } else {
+        CHECK(false) << "should not get here";
+      }
+      break;
+
+    case runtime::DataType::TypeCode::kFloat: {
+      os.fill(' ');
+      os.setf(std::ios::left, std::ios::adjustfield);
+      if (arr_type.bits() == 32) {
+        PrintArray<float>(tensor->dl_tensor.data, num_elements, one_element_size_bytes,
```

Comment:
I still think it is easier to use uint to print the case of float in most of our cases. It also helps us to get around the problem of printing exp and mantisa

Reply:
I mean, the printing problem is just an alignment thing. it doesn't affect the functionality. I don't see why adding an implicit cast to generated code that the compiler will never even know about is easier? because remember, we are not going to add the cast into the generated function that is accessing these floats--it is just going to get a float* and then by generating uint32_t here, the float* will magically access uint32_t data. I agree with you that it doesn't seeeem like anything would break too badly given they are the same size, but it's kind of a long limb to go out on just to support bfloat16. we would also need to do byte ordering.

Reply:
To put it in another way. We will need support for bfloat16 and fp16 now given that they are already part of the type system. Addtiionally, there might also be need to support customized data types. 

The main reason to go for the uint repr is its bit accurate precision(we don't need to worry about printing out INF, nan, or loss of accuracy due to printing and parsing of floating pt -- although partly addressed by hex printing, still somewhat complicated) and simplicity in implementation.

Additionally, for big endian machines and small endian machines we are already using the bits as the only information in our runtime https://github.com/apache/tvm/blob/main/include/tvm/runtime/ndarray.h#L429, which means our runtime won't simply work beyond such cases. Right now we also did not see an example that goes beyond these two cases.

If we do think it is important to keep float and double printing as their own type. I think we could do that, but need to make sure the rest of the codepath uses uint so that it works for bfloat16, fp16 and other potential extension data types now.

Reply:
okay, as discussed i've added support for float16 and bfloat16. left the rest of the float printing as it was before.